rfcs/2023-07-03-theme-switching-ssr.md

```diff
@@ -0,0 +1,124 @@
+
+This approach is the bulk of changes that would be required on Blade side, following sub-sections cover details on achieving the same.
+
+### Generate CSS variables
+
+We're already generating these in `generateCssThemeVariables.js`. Build outputs different CSS files for payment and banking themes in light and dark modes. However this requires users to include the CSS files on their sides.
+
+Since most users don't need to include the CSS theme files, we can populate these color based CSS variables from `ThemeProvider` using [global styles](https://styled-components.com/docs/api#createglobalstyle). With this, the CSS variables will be available in `:root` context.
+
+### Changes for reading color tokens in Blade
+
+Writing `var(--colors-surface-action-icon-hover-lowContrast)` like syntax every time is cumbersome and error prone. Instead we can add a utility that reads color tokens from CSS variables:
+
+```tsx
+//...
+// returns the CSS variable: var(--colors-surface-action-icon-hover-lowContrast)
+color: getColorFromCSSVariables('surface.action.icon.hover.lowContrast');
+```
+
+To keep parity with native, the `.native` version of this utility would operate on `theme` so `.native` modules will continue to work as it is.
+
+In order to prevent us from reading `theme.colors` directly after this change, we can add a [linting rule](https://eslint.org/docs/latest/rules/no-restricted-properties). This will flag occurrences wherever we accidentally read `theme.colors`.
+
+### Script utilities
+
+- `getInitialColorMode` - a function that would read user's preference from either `localStorage` (if present) or else fallback to `matchMedia` `prefers-color-scheme` and return `light` or `dark`. This will be used by `ThemeProvider` and may be used by user's for any custom logic.
+- `BLADE_COLOR_MODE_INLINE_SCRIPT` - a serialized script that users can put in their HTML, this will be blocking and update CSS variables based on user preferences (light / dark).
+- Setter and getter utilities - for reading and writing color modes, this will be useful to build color scheme toggles by users - these will involve syncing color mode preference to and from `localStorage` and also updating required CSS variables.
+
+## Drawbacks/Constraints
+
+Primary drawback for this approach is increase in complexity on Blade side, however the tradeoff is essential for creating a seamless theme switching UX for SSR products.
+
+## Alternatives
+
+- `Sec-CH-Prefers-Color-Scheme` header [reference](https://web.dev/user-preference-media-features-headers/#demo-of-sec-ch-prefers-color-scheme) to make server have context of user preferences and accordingly send an appropriate serialized markup. However [browser support](https://caniuse.com/?search=Sec-CH-Prefers-Color-Scheme) is limited (recent Chrome versions only).
```

Comment:
can we check for docs and website what is the % split for our users across browsers?

Reply:
+1